x/poolmanager/routes.go

```diff
@@ -0,0 +1,597 @@
+					clonedMiddleRoute := *middleRoute
+					clonedMiddleRoute.Token = token2
+
+					for _, endRoute := range endRoutes.Routes {
+						clonedEndRoute := *endRoute
+						clonedEndRoute.Token = end
+
+						routeRoutes = append(routeRoutes, []*types.Route{&clonedStartRoute, &clonedMiddleRoute, &clonedEndRoute})
+					}
+				}
+			}
+		}
+	}
+
+	return routeRoutes
+}
+
+// SetDenomPairRoutes sets the route map to be used for route calculations
+func (k Keeper) SetDenomPairRoutes(ctx sdk.Context) (types.RoutingGraph, error) {
+	// Get all the pools
+	pools, err := k.AllPools(ctx)
+	if err != nil {
+		return types.RoutingGraph{}, err
+	}
+
+	// Create a routingGraph to represent possible routes between tokens
+	var routingGraph types.RoutingGraph
+
+	// Iterate through the pools
+	for _, pool := range pools {
+		tokens := pool.GetPoolDenoms(ctx)
+		poolID := pool.GetId()
+		// Create edges for all possible combinations of tokens
+		for i := 0; i < len(tokens); i++ {
+			for j := i + 1; j < len(tokens); j++ {
+				// Add edges with the associated token
+				routingGraph.AddEdge(tokens[i], tokens[j], tokens[i], poolID)
+				routingGraph.AddEdge(tokens[j], tokens[i], tokens[j], poolID)
+			}
+		}
+	}
+
+	// Set the route map in state
+	// NOTE: This is done with the non map version of the route graph
+	// If we used maps here, the serialization would be non-deterministic
+	osmoutils.MustSet(ctx.KVStore(k.storeKey), types.KeyRouteMap, &routingGraph)
+	return routingGraph, nil
+}
+
+// GetDenomPairRoute returns the route with the lowest slippage between an input denom and an output denom
+// The method starts by finding all direct routes, two hop, and three hop routes.
+// It then calculates the liquidity of each route and selects the route with the highest liquidity for each hop count.
+// In other words, at a max, there will be one direct route, one two hop route, and one three hop route.
+// The method then simulates a swap against each route and selects the route with the highest swap amount out.
+func (k Keeper) GetDenomPairRoute(ctx sdk.Context, inputCoin sdk.Coin, outputDenom string) ([]types.SwapAmountInRoute, error) {
+	inputDenom := inputCoin.Denom
+
+	// Ensure the caches are restored when the function exits
+	defer func() {
+		directRouteCache = make(map[string]uint64)
+		spotPriceCache = make(map[string]osmomath.BigDec)
+	}()
+
+	// Retrieve the route map from state
+	routeMap, err := k.GetRouteMap(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get all direct routes, two hop, and three hop routes
+	directPoolIDs := findDirectRoute(routeMap, inputDenom, outputDenom)
+
+	var twoHopPoolIDs [][]*types.Route
+	if inputDenom != OSMO && outputDenom != OSMO {
+		twoHopPoolIDs = findTwoHopRoute(routeMap, inputDenom, outputDenom)
+	}
+
+	var threeHopPoolIDs [][]*types.Route
+	if inputDenom != OSMO && outputDenom != OSMO {
+		threeHopPoolIDs = findThreeHopRoute(routeMap, inputDenom, outputDenom)
+	}
```

Comment:
Can this be done iteratively? What is the difference in handling each subsequent hop?

Reply:
I think this is what you were asking to be done https://github.com/osmosis-labs/osmosis/pull/6895/commits/9676acae390e1f68253fc8966eeafd0c8242115c